File: src/common/py/core/messaging/dispatchers/command_reply_dispatcher.py
from .message_dispatcher import MessageDispatcher
from ..command_reply import CommandReply
from ..meta import CommandReplyMetaInformation


class CommandReplyDispatcher(MessageDispatcher[CommandReply]):
    # pylint: disable=protected-access
    """
    Message dispatcher specific to ``CommandReply``.
    """

    def __init__(self):
        super().__init__(CommandReplyMetaInformation)

    def pre_dispatch(
        self, msg: CommandReply, msg_meta: CommandReplyMetaInformation
    ) -> None:
        """
        Invokes reply callbacks associated with the replied command.

        Args:
            msg: The command reply that is about to be dispatched.
            msg_meta: The command reply meta information.
        """
        from .command_dispatcher import CommandDispatcher
        from ...logging import debug

        debug(f"Dispatching command reply: {msg}", scope="bus")
        super().pre_dispatch(msg, msg_meta)

        CommandDispatcher.invoke_reply_callbacks(msg.unique, reply=msg)
        MessageDispatcher._meta_information_list.remove(msg.unique)

File: src/common/web/core/messaging/dispatchers/CommandReplyDispatcher.ts
import logging from "../../logging/Logging";
import { CommandReply } from "../CommandReply";
import { CommandReplyMetaInformation } from "../meta/CommandReplyMetaInformation";
import { CommandDispatcher } from "./CommandDispatcher";
import { MessageDispatcher } from "./MessageDispatcher";

/**
 * Message dispatcher specific to ``CommandReply``.
 */
export class CommandReplyDispatcher extends MessageDispatcher<CommandReply, CommandReplyMetaInformation> {
    /**
     * Called to perform tasks *before* sending a message.
     *
     * This method is called before any service-registered message handler is invoked.
     *
     * @param msg - The message that is about to be dispatched.
     * @param msgMeta - The message meta information.
     *
     * @throws Error - If the meta information type is invalid.
     */
    public preDispatch<MsgType extends CommandReply>(msg: MsgType, msgMeta: CommandReplyMetaInformation): void {
        logging.debug(`Dispatching command reply: ${String(msg)}`, "bus");
        super.preDispatch(msg, msgMeta);

        CommandDispatcher.invokeReplyCallbacks(msg.unique, msg, msgMeta);
        MessageDispatcher._metaInformationList.remove(msg.unique);
    }
}
